weaviate/connect/auth.py

```diff
@@ -0,0 +1,217 @@
+from weaviate.warnings import _Warnings
+
+AUTH_DEFAULT_TIMEOUT = 5
+OIDC_CONFIG = Dict[str, Union[str, List[str]]]
+
+
+class _Auth:
+    def __init__(
+        self,
+        oidc_config: OIDC_CONFIG,
+        credentials: AuthCredentials,
+        connection: _ConnectionBase,
+    ) -> None:
+        self._credentials: AuthCredentials = credentials
+        self._connection: _ConnectionBase = connection
+        config_url = oidc_config["href"]
+        client_id = oidc_config["clientId"]
+        assert isinstance(config_url, str) and isinstance(client_id, str)
+        self._open_id_config_url: str = config_url
+        self._client_id: str = client_id
+        self._default_scopes: List[str] = []
+        if "scopes" in oidc_config:
+            default_scopes = oidc_config["scopes"]
+            assert isinstance(default_scopes, list)
+            self._default_scopes = default_scopes
+
+        self._token_endpoint: str = self._get_token_endpoint()
+        self._validate(oidc_config)
+
+    def _validate(self, oidc_config: OIDC_CONFIG) -> None:
+        if isinstance(self._credentials, AuthClientPassword):
+            if self._token_endpoint.startswith("https://login.microsoftonline.com"):
+                raise AuthenticationFailedException(
+                    """Microsoft/azure does not recommend to authenticate using username and password and this method is
+                    not supported by the python client."""
+                )
+
+            # The grant_types_supported field is optional and does not have to be present in the response
+            if (
+                "grant_types_supported" in oidc_config
+                and "password" not in oidc_config["grant_types_supported"]
+            ):
+                raise AuthenticationFailedException(
+                    """The grant_types supported by the third-party authentication service are insufficient. Please add
+                    the 'password' grant type."""
+                )
+
+    def _get_token_endpoint(self) -> str:
+        response_auth = get(self._open_id_config_url, proxies=self._connection.get_proxies())
+        response_auth_json = _decode_json_response_dict(response_auth, "Get token endpoint")
+        assert response_auth_json is not None
+        token_endpoint = response_auth_json["token_endpoint"]
+        assert isinstance(token_endpoint, str)
+        return token_endpoint
+
+    def get_sync_auth_client(self) -> OAuth2Client:
+        if isinstance(self._credentials, AuthBearerToken):
+            session = self._get_client_auth_bearer_token(self._credentials)
+        elif isinstance(self._credentials, AuthClientCredentials):
+            session = self._get_client_client_credential(self._credentials)
+        else:
+            assert isinstance(self._credentials, AuthClientPassword)
+            session = self._get_client_user_pw(self._credentials)
+
+        return session
+
+    async def get_async_auth_client(self) -> AsyncOAuth2Client:
```

Comment:
this is currently not used, right? There can always be only one oidc-auth connection

Reply:
Yes, will remove this file as above